File: Queue.cs
namespace Queues
{
    class Queue<T>
    {
        public Node<T> Head
        { get; set; }
        public Node<T> Tail
        { get; set; }
        public int MaxSize
        { get; }
        public int Size
        { get; private set; }

        public Queue(int? maxSize = null)
        {
            Head = null;
            Tail = null;
            MaxSize = (int)maxSize;
            Size = 0;
        }
    }
}

File: queue.rb
require './node.rb'

class Queue
    attr_reader :max_size
    attr_accessor :head, :tail, :size

    def initialize(max_size = nil)
        @head = nil
        @tail = nil
        @max_size = max_size
        @size = 0
    end

    def has_space()
        if @max_size == nil
            return true
        else
            return @max_size > @size
        end
    end

    def is_empty()
        @size == 0
    end

end